plotly/tools.py

```diff
@@ -2871,3 +3061,152 @@ def get_candle_decrease(self):
+            self.colors = kwargs['colors']
+        else:
+            self.colors = [
+                "rgb(31, 119, 180)", "rgb(255, 127, 14)",
+                "rgb(44, 160, 44)", "rgb(214, 39, 40)",
+                "rgb(148, 103, 189)", "rgb(140, 86, 75)",
+                "rgb(227, 119, 194)", "rgb(127, 127, 127)",
+                "rgb(188, 189, 34)", "rgb(23, 190, 207)"]
+        self.curve_x = [None] * self.trace_number
+        self.curve_y = [None] * self.trace_number
+
+        for trace in self.hist_data:
+            self.start.append(min(trace) * 1.)
+            self.end.append(max(trace) * 1.)
+
+    def make_hist(self):
+        """
+        Makes the histogram(s) for FigureFactory.create_distplot().
+
+        :rtype (list) hist: list of histogram representations
+        """
+        hist = [None] * self.trace_number
+
+        for index in range(self.trace_number):
+            hist[index] = dict(type='histogram',
+                               x=self.hist_data[index],
+                               xaxis='x1',
+                               yaxis='y1',
+                               histnorm='probability',
+                               name=self.group_labels[index],
+                               legendgroup=self.group_labels[index],
+                               marker=dict(color=self.colors[index]),
+                               autobinx=False,
+                               xbins=dict(start=self.start[index],
+                                          end=self.end[index],
+                                          size=self.bin_size),
+                               opacity=.7)
+        return hist
+
+    def make_kde(self):
+        """
+        Makes the kernal density estimation(s) for create_distplot().
+
+        This is called when curve_type = 'kde' in create_distplot().
+
+        :rtype (list) curve: list of kde representations
+        """
+        curve = [None] * self.trace_number
+        for index in range(self.trace_number):
+            self.curve_x[index] = [self.start[index] +
+                                   x * (self.end[index] - self.start[index])
+                                   / 500 for x in range(500)]
+            self.curve_y[index] = (scipy.stats.gaussian_kde
+                                   (self.hist_data[index])
+                                   (self.curve_x[index]))
+            self.curve_y[index] *= self.bin_size
+
+        for index in range(self.trace_number):
+            curve[index] = dict(type='scatter',
+                                x=self.curve_x[index],
+                                y=self.curve_y[index],
+                                xaxis='x1',
+                                yaxis='y1',
+                                mode='lines',
+                                name=self.group_labels[index],
+                                legendgroup=self.group_labels[index],
+                                showlegend=False,
```

Comment:
funny side affect. if `show_hist` is `False` then no legend appears. So, we'll have to set `showlegend=True` to the histogram if it's shown, otherwise the KDE if it's shown, otherwise the rug if it's shown

![image](https://cloud.githubusercontent.com/assets/1280389/9557278/b5cd5662-4da8-11e5-80c1-4c35879adb48.png)